Python/chapter03/p03_stack_of_plates/miguelHx.py

```diff
@@ -0,0 +1,285 @@
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty.')
+        item = self.top.data
+        self.top = self.top.next
+        self.current_node = self.top
+        self.size -= 1
+        return item
+
+    def push(self, item: T) -> None:
+        """
+        Adds an item to the top of the stack
+        Args:
+            item (int): data we want at the top of stack
+        """
+        t = StackNode(item, None)
+        t.next = self.top
+        self.top = t
+        self.current_node = self.top
+        self.size += 1
+
+    def peek(self) -> T:
+        """
+        Returns data at the top of the stack
+        Raises:
+            IndexError: [description]
+        Returns:
+            int: the value at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty')
+        return self.top.data
+
+    def __iter__(self) -> Iterator:
+        """
+        Builds a list of the current stack state.
+        For example, given the following stack:
+            3 -> 2 -> 1, where 3 is the top,
+        Expect:
+            [3, 2, 1]
+        Returns:
+            List[int]: list of integers
+        """
+        return self
+
+    def __next__(self) -> T:
+        self.index += 1
+        if self.index == self.size or self.current_node is None:
+            self.index = -1
+            self.current_node = self.top
+            raise StopIteration
+        n: T = self.current_node.data
+        self.current_node = self.current_node.next
+        return n
+
+    def __bool__(self) -> bool:
+        """
+        True is returned when the container is not empty.
+        From https://docs.python.org/3/reference/datamodel.html#object.__bool__ :
+        Called to implement truth value testing and the built-in operation bool();
+        should return False or True. When this method is not defined, len() is called,
+        if it is defined, and the object is considered true if its result is nonzero.
+        If a class defines neither len() nor bool(), all its instances are considered true.
+        Returns:
+            bool: False when empty, True otherwise
+        """
+        return self.size > 0
+    
+    def __len__(self) -> int:
+        return self.size
+
+    def __str__(self):
+        if self.size == 0:
+            return '<Empty>'
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(str(n.data))
+            n = n.next
+        values.append(str(n.data))
+        return '->'.join(values)
+
+
+
+class SetofStacks:
+
+    def __init__(self):
+        self.current_stack_idx: int = 0
+        self.set_of_stacks: List[MyStack] = [MyStack()]
+        self.stack_threshold: int = 3
+        self.size: int = 0
+    
+    def push(self, item: T) -> None:
+        # threshold check
+        if len(self.set_of_stacks[self.current_stack_idx]) >= self.stack_threshold:
```

Comment:
optional:

```suggestion
        if len(self.set_of_stacks[-1]) >= self.stack_threshold:
```

as long as the length of `set_of_stacks` is kept to be at least 1, `-1` as an index will [always point to the last element](https://pythonguides.com/indexing-and-slicing-in-python/#Negative_indexing_example_in_Python). no need to keep `current_stack_idx` around!

Reply:
Suggestion included along with other updates here:  c60ef5551631b4d1fd880ff2e066e0726ca1410c